File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums):
        nums.sort()
        length = len(nums)
        answerset = set()
        answer = []
        for i in range(length -2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            b,e = i+1 , length-1
            while b < e :
                sums = nums[i]+nums[b]+nums[e]
                if sums > 0:
                    e-=1
                elif sums < 0:
                    b+=1
                else:
                    answer.append([nums[i],nums[b],nums[e]])
                    b+=1
                    e-=1
                    while b < e and nums[b] == nums[b - 1]:
                        b += 1
                    while b < e and nums[e] == nums[e + 1]:
                        e -= 1
                
        return answer 

File: 15-3sum/3sum.cs
public class Solution {
    public IList<IList<int>> ThreeSum(int[] nums) {
        Array.Sort(nums);
        IList<IList<int>> answer = new List<IList<int>>();
        for(int i=0 ; i < nums.Length - 2; i++){
           if (i > 0 && nums[i] == nums[i - 1])
            {
                continue;
            }
           int middle = i+1;
           int end = nums.Length-1 ;
           while(end > middle )
           {
            int sum = nums[i] + nums[middle]+ nums[end];
            if(sum == 0){
                answer.Add(new List<int>{nums[i],nums[middle],nums[end]});
                end--;
                while( end > middle && nums[end] == nums[end+1]){
                    end-=1;
                }
                middle+=1;
                while( middle < end && nums[middle] == nums[middle-1]){
                    middle+=1;
                }
            }
            else if (sum > 0){
                end -=1;
            }
            else{
                middle +=1;
            }

           }

        }
        return answer ;
    }
}